[css-anchor-position-1] Chained anchors don't scroll with their anchor

anttijk opened the pull request:
#### 6e090026a7cd2c31b7b9c9d13cbe964d56569040
<pre>
[css-anchor-position-1] Chained anchors don&apos;t scroll with their anchor
<a href="https://bugs.webkit.org/show_bug.cgi?id=295012">https://bugs.webkit.org/show_bug.cgi?id=295012</a>
<a href="https://rdar.apple.com/154355127">rdar://154355127</a>

Reviewed by Alan Baradlay.

* LayoutTests/TestExpectations:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::scrollOffsetFromAnchor):

The anchor may be scroll-compensated itself. Include these offsets recursively.
Recursion ends because anchors can&apos;t form cycles.

Canonical link: <a href="https://commits.webkit.org/297286@main">https://commits.webkit.org/297286@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c550dd0b3c320a351d59a803f1a9d26870f7da3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111162 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117193 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61430 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113124 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39410 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84503 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114109 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25164 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100100 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64949 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24509 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/18242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61013 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94544 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18309 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120195 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38211 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28396 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93439 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-chained-002.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38587 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96376 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93264 "Found 1 new API test failure: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23767 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38349 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16119 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34196 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38100 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43577 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37765 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41098 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39467 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7a272c913e63ab7eb592f3943e26c29b6bc0fcc4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47722

fantasai commented:
The merge conflict is super basic -- we're passing the anchor-scope-scroll.html test now, so it's been removed from TestExpectations, but that messes up the context for this patch which removes failures for two other tests.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/62fa0b0dc84f18a911b3f76978b8faf4e7524242. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47722

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2c550dd0b3c320a351d59a803f1a9d26870f7da3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47722

anttijk committed:
[css-anchor-position-1] Chained anchors don't scroll with their anchor
https://bugs.webkit.org/show_bug.cgi?id=295012
rdar://154355127

Reviewed by Alan Baradlay.

* LayoutTests/TestExpectations:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::scrollOffsetFromAnchor):

The anchor may be scroll-compensated itself. Include these offsets recursively.
Recursion ends because anchors can't form cycles.

Canonical link: https://commits.webkit.org/297286@main
